daffodil-core/src/main/scala/org/apache/daffodil/compiler/Compiler.scala

```diff
@@ -73,190 +73,169 @@ object ForParser extends ParserOrUnparser
-      val p = if (!rootElem.isError) parser else null
-      val u = if (!rootElem.isError) unparser else null
-      val ssrd = new SchemaSetRuntimeData(
-        p,
-        u,
-        this.diagnostics,
-        rootERD,
-        variables,
-        validationMode,
-        sset.typeCalcMap)
-      if (rootElem.numComponents > rootElem.numUniqueComponents)
-        log(LogLevel.Info, "Compiler: component counts: unique %s, actual %s.",
-          rootElem.numUniqueComponents, rootElem.numComponents)
-      val dataProc = new DataProcessor(ssrd, tunable)
-      if (dataProc.isError) {
-        // NO longer printing anything here. Callers must do this.
-        //        val diags = dataProc.getDiagnostics
-        //        log(LogLevel.Error,"Compilation (DataProcessor) reports %s compile errors/warnings.", diags.length)
-        //        diags.foreach { diag => log(LogLevel.Error, diag.toString()) }
-      } else {
-        log(LogLevel.Compile, "Parser = %s.", ssrd.parser.toString)
-        //log(LogLevel.Error, "Unparser = %s.", ssrd.unparser.toString)
-        log(LogLevel.Compile, "Compilation (DataProcesor) completed with no errors.")
-      }
-      dataProc
-    }
-  }
-}
+  private def this(validateDFDLSchemas: Boolean = true) = this(validateDFDLSchemas, DaffodilTunables(), Queue.empty, true, None, None)
 
-/**
- * Both Compiler and ProcessorFactory share this same API call.
- */
-trait HavingRootSpec extends Logging {
-  var rootSpec: Option[RootSpec] = None
+  private def copy(validateDFDLSchemas: Boolean = validateDFDLSchemas,
+    tunables : DaffodilTunables = tunables,
+    externalDFDLVariables: Queue[Binding] = externalDFDLVariables,
+    checkAllTopLevel : Boolean = checkAllTopLevel,
+    optRootName: Option[String] = optRootName,
+    optRootNamespace: Option[String] = optRootNamespace) =
+    new Compiler(validateDFDLSchemas, tunables, externalDFDLVariables, checkAllTopLevel, optRootName, optRootNamespace)
 
+  @deprecated("Pass argument to compileSource, or compileFile.", "Since 2.6.0")
   def setDistinguishedRootNode(name: String, namespace: String): Unit = {
-    val ns =
-      if (namespace != null) Some(NS(namespace))
-      else None
-    rootSpec = Some(RootSpec(ns, name))
-    // log(Info("%s setDistinguishedRootNode to %s", Misc.getNameFromClass(this), rootSpec))
-    //
-    // null means we search for the namespace
-    // Must be only one answer.
-    //
-
+    Assert.usage(name ne null)
+    optRootName = Option(name)
+    optRootNamespace = Option(namespace)
   }
-}
-class InvalidParserException(msg: String, cause: Throwable = null) extends Exception(msg, cause)
 
-class Compiler(var validateDFDLSchemas: Boolean = true)
-  extends DFDL.Compiler
-  with Logging
-  with HavingRootSpec {
+  def withDistinguishedRootNode(name: String, namespace: String) : Compiler = {
+    Assert.usage(name ne null)
+    copy(optRootName = Option(name), optRootNamespace = Option(namespace))
+  }
```

Comment:
You have to have this in order to support the old API style (which is still in place in JAPI/SAPI, and the TDML Processor), without having to call a deprecated method, which you can't do without us turning off -Xfatal-warnings at compilation time. So even though we don't technically want this method, there's nothing wrong with it per-se. It's functional anyway. 

I'd rather not have it, just because I'd like there to ultimately be one way to do things, and the best functional interface is one function with a bunch of optional args.  

But, I think we have to have this or we have to convert JAPI/SAPI and the TDML processor all at the same time, which is, frankly, too much change all at once, and the implications of it are not clear.  Even if we wanted to change the JAPI/SAPI and TDML processor APIs, we still have to support their deprecated behavior for now, so we'd *still* need this "with-X" method for the time being.

Reply:
I have updated everything now to deprecate all setters in JAPI/SAPI, and TDML processor.

Reply:
btw: this method on compiler withDistinguishedRootNode is needed to implement the deprecated setter in JAPI/SAPI, and the TDML processor. It's not part of the API officially. Not in DFDL.Compiler.